runtime/gc_ctrl.c

```diff
@@ -297,7 +296,16 @@ CAMLprim value caml_gc_compaction(value v)
   Caml_check_caml_state();
   CAML_EV_BEGIN(EV_EXPLICIT_GC_COMPACT);
   CAMLassert (v == Val_unit);
-  value exn = gc_major_exn(1);
+  value exn = Val_unit;
+  int i;
+  /* We do a full major before this compaction.
+     See [caml_full_major_exn] for this needs
```

Comment:
for +why+ this needs three iterations?

Reply:
Oops, fixed in https://github.com/ocaml/ocaml/pull/12859/commits/dae4b1f128796ab17fc001ab0792edb284e92037